File: Android/TUIKit/TUICallKit/tuicallkit/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/utils/ImageLoader.java
package com.tencent.qcloud.tuikit.tuicallkit.utils;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapShader;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.drawable.Drawable;
import android.widget.ImageView;

import androidx.annotation.DrawableRes;
import androidx.annotation.RawRes;

import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestBuilder;
import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool;
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation;
import com.bumptech.glide.request.RequestOptions;
import com.tencent.qcloud.tuikit.tuicallkit.R;

import java.security.MessageDigest;
import java.util.concurrent.ExecutionException;

public class ImageLoader {
    private static int radius = 0;

    public static void clear(Context context, ImageView imageView) {
        Glide.with(context).clear(imageView);
    }

    public static void loadImage(Context context, ImageView imageView, Object url) {
        loadImage(context, imageView, url, R.drawable.tuicalling_ic_avatar);
    }

    public static void loadImage(Context context, ImageView imageView, Object url, @DrawableRes int resId) {
        if (url == null || url == "") {
            if (imageView != null && resId != 0) {
                imageView.setImageResource(resId);
            }
            return;
        }
        Glide.with(context).load(url).placeholder(resId).error(loadTransform(context, resId, radius)).into(imageView);
    }

    public static void loadImage(Context context, ImageView imageView, @RawRes @DrawableRes Integer resourceId) {
        Glide.with(context).load(resourceId).into(imageView);
    }

    public static Bitmap loadBitmap(Context context, Object imageUrl, int targetImageSize) throws InterruptedException,
            ExecutionException {
        if (imageUrl == null) {
            return null;
        }
        return Glide.with(context).asBitmap().load(imageUrl)
                .apply(loadTransform(context, R.drawable.tuicalling_ic_avatar, radius))
                .into(targetImageSize, targetImageSize)
                .get();
    }

    public static void loadGifImage(Context context, ImageView imageView, @RawRes @DrawableRes int resourceId) {
        Glide.with(context).asGif().load(resourceId).into(imageView);
    }

    private static RequestBuilder<Drawable> loadTransform(Context context, @DrawableRes int resId, int radius) {
        return Glide.with(context).load(resId).placeholder(resId)
                .apply(new RequestOptions().centerCrop().transform(new GlideRoundTransform(context, radius)));
    }

    public static class GlideRoundTransform extends BitmapTransformation {
        private static float radius = 0f;

        public GlideRoundTransform(Context context, int dp) {
            radius = Resources.getSystem().getDisplayMetrics().density * dp;
        }

        @Override
        protected Bitmap transform(BitmapPool pool, Bitmap toTransform, int outWidth, int outHeight) {
            return roundCrop(pool, toTransform);
        }

        private static Bitmap roundCrop(BitmapPool pool, Bitmap source) {
            if (source == null) {
                return null;
            }

            Bitmap result = pool.get(source.getWidth(), source.getHeight(), Bitmap.Config.ARGB_8888);
            if (result == null) {
                result = Bitmap.createBitmap(source.getWidth(), source.getHeight(), Bitmap.Config.ARGB_8888);
            }

            Canvas canvas = new Canvas(result);
            Paint paint = new Paint();
            paint.setShader(new BitmapShader(source, BitmapShader.TileMode.CLAMP, BitmapShader.TileMode.CLAMP));
            paint.setAntiAlias(true);
            RectF rectF = new RectF(0f, 0f, source.getWidth(), source.getHeight());
            canvas.drawRoundRect(rectF, radius, radius, paint);
            return result;
        }

        @Override
        public void updateDiskCacheKey(MessageDigest messageDigest) {

        }
    }
}

File: Android/TUIKit/atomic_x/src/main/java/io/trtc/tuikit/atomicx/widget/utils/ImageLoader.kt
package io.trtc.tuikit.atomicx.widget.utils

import android.app.Activity
import android.content.Context
import android.graphics.Bitmap
import android.graphics.RenderEffect
import android.graphics.Shader
import android.os.Build
import android.util.TypedValue
import android.widget.ImageView
import androidx.annotation.DrawableRes
import androidx.annotation.RawRes
import com.bumptech.glide.Glide
import com.bumptech.glide.RequestBuilder
import com.bumptech.glide.load.MultiTransformation
import com.bumptech.glide.load.Transformation
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation
import com.bumptech.glide.load.resource.bitmap.CenterCrop
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.bumptech.glide.request.RequestOptions
import java.security.MessageDigest

object ImageLoader {

    @JvmStatic
    fun load(context: Context, target: ImageView?, source: Any?, @DrawableRes placeImage: Int) {
        val imageOptions = ImageOptions.Builder()
            .setPlaceImage(placeImage)
            .build()
        load(context, target, source, imageOptions)
    }

    @JvmStatic
    fun load(context: Context, target: ImageView?, source: Any?, config: ImageOptions) {
        if (target == null) {
            return
        }

        if (source == null) {
            val image = if (config.placeImage != 0) config.placeImage else config.errorImage
            target.setImageResource(image)
            return
        }

        loadImageView(context, target, source, config)
    }

    @JvmStatic
    fun loadGif(context: Context, target: ImageView?, @RawRes @DrawableRes resourceId: Int) {
        if (target == null) {
            return
        }
        Glide.with(context.applicationContext)
            .asGif()
            .load(resourceId)
            .into(target)
    }

    @JvmStatic
    fun transformBitmap(
        context: Context,
        source: Any?,
        width: Int,
        height: Int,
        options: ImageOptions
    ): Bitmap? {
        if (source == null) {
            return null
        }

        val builder = Glide.with(context.applicationContext)
            .asBitmap()
            .load(source)

        setBuilderOptions(context.applicationContext, builder, options)

        return try {
            builder.submit(width, height).get()
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    @JvmStatic
    fun clear(context: Context, target: ImageView?) {
        if (target != null) {
            Glide.with(context.applicationContext).clear(target)
        }
    }

    private fun loadImageView(context: Context, target: ImageView, source: Any, config: ImageOptions) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            if (context is Activity) {
                if (context.isFinishing || context.isDestroyed) {
                    return
                }
            }
        }

        val builder = if (config.isGif) {
            Glide.with(context.applicationContext).asGif().load(source)
        } else {
            Glide.with(context.applicationContext).load(source)
        }

        setBuilderOptions(context.applicationContext, builder, config)
        builder.into(target)
        setRenderEffect(target, config)
    }

    private fun setRenderEffect(target: ImageView, config: ImageOptions) {
        val level = config.blurEffect
        if (level > 0 && Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            target.setRenderEffect(
                RenderEffect.createBlurEffect(level, level, Shader.TileMode.MIRROR)
            )
        }
    }

    private fun setBuilderOptions(context: Context, builder: RequestBuilder<*>, config: ImageOptions) {
        val options = RequestOptions()

        val transformations = mutableListOf<Transformation<Bitmap>>()
        transformations.add(CenterCrop())

        if (config.roundRadius > 0) {
            transformations.add(RoundedCorners(dpToPx(context, config.roundRadius)))
        }

        if (config.blurEffect > 0 && Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            transformations.add(BlurTransformation(context, config.blurEffect))
        }

        options.transform(MultiTransformation(transformations))

        if (config.placeImage != 0) {
            options.placeholder(config.placeImage)
        }

        if (config.errorImage != 0) {
            options.error(config.errorImage)
        }

        options.diskCacheStrategy(
            if (config.skipDiskCache) DiskCacheStrategy.NONE else DiskCacheStrategy.ALL
        )
        options.skipMemoryCache(config.skipMemoryCache)

        builder.apply(options)
    }

    private fun dpToPx(context: Context, dp: Int): Int {
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP,
            dp.toFloat(),
            context.resources.displayMetrics
        ).toInt()
    }

    class BlurTransformation(
        context: Context,
        level: Float
    ) : BitmapTransformation() {

        private val radius: Float = level * 0.25f
        private val contextRef = java.lang.ref.WeakReference(context)

        override fun updateDiskCacheKey(messageDigest: MessageDigest) {
            messageDigest.update("blur:$radius".toByteArray())
        }

        override fun transform(
            pool: BitmapPool,
            toTransform: Bitmap,
            outWidth: Int,
            outHeight: Int
        ): Bitmap {
            val context = contextRef.get() ?: return toTransform
            return BlurUtils.blur(context, pool, toTransform, radius.toInt())
        }
    }
}
